indexer/src/main/java/au/org/aodn/esindexer/service/VocabServiceImpl.java

```diff
@@ -0,0 +1,547 @@
+    protected Function<JsonNode, String> definition = (node) -> node.has("definition") ? node.get("definition").asText() : null;
+    protected BiFunction<JsonNode, String, Boolean> isNodeValid = (node, item) -> node != null && !node.isEmpty() && node.has(item) && !node.get(item).isEmpty();
+
+    private VocabModel buildVocabByResourceUri(String vocabUri, String vocabApiBase, VocabApiPaths vocabApiPaths) {
+        String resourceDetailsApi = vocabUri.contains("_classes")
+                ? vocabApiPaths.getVocabCategoryDetailsApiPath()
+                : vocabApiPaths.getVocabDetailsApiPath();
+
+        String detailsUrl = String.format(vocabApiBase + resourceDetailsApi, vocabUri);
+
+        try {
+            log.debug("Query api -> {}", detailsUrl);
+            ObjectNode detailsObj = indexerRestTemplate.getForObject(detailsUrl, ObjectNode.class);
+            if(isNodeValid.apply(detailsObj, "result") && isNodeValid.apply(detailsObj.get("result"), "primaryTopic")) {
+                JsonNode target = detailsObj.get("result").get("primaryTopic");
+
+                VocabModel vocab = VocabModel
+                        .builder()
+                        .label(label.apply(target))
+                        .definition(definition.apply(target))
+                        .about(vocabUri)
+                        .build();
+
+                List<VocabModel> narrowerNodes = new ArrayList<>();
+                if (isNodeValid.apply(target, "narrower")) {
+                    for (JsonNode j : target.get("narrower")) {
+                        if (!about.apply(j).isEmpty()) {
+                            // recursive call
+                            VocabModel narrowerNode = buildVocabByResourceUri(about.apply(j), vocabApiBase, vocabApiPaths);
+                            if (narrowerNode != null) {
+                                narrowerNodes.add(narrowerNode);
+                            }
+                        }
+                    }
+                }
+
+                if (!narrowerNodes.isEmpty()) {
+                    vocab.setNarrower(narrowerNodes);
+                }
+
+                return vocab;
+            }
+        } catch(Exception e) {
+            log.error("Item not found in resource {}", detailsUrl);
+        }
+        return null;
+    }
+
+    protected <T> VocabModel buildVocabModel(T currentNode, String vocabApiBase, VocabApiPaths vocabApiPaths) {
+        String resourceUri = null;
+
+        if (currentNode instanceof ObjectNode objectNode) {
+            resourceUri = objectNode.has("_about") ? about.apply(objectNode) : objectNode.asText();
+        } else if (currentNode instanceof TextNode textNode) {
+            resourceUri = textNode.asText();
+        } else if (currentNode instanceof VocabModel vocabNode) {
+            String about = vocabNode.getAbout();
+            if (about != null && !about.isEmpty()) {
+                resourceUri = about;
+            }
```

Comment:
risk getting null pointer exception here, I'll change to

```java
            if (vocabNode.getAbout() != null && !vocabNode.getAbout().isEmpty()) {
                resourceUri = vocabNode.getAbout();
            }
```

Reply:
use safeGet() as much as you can it is easier

Reply:
will do altogether after the PR view